### ISSUE TITLE
Signal handler doesn't catch cases where the stack is exhausted

### ISSUE THREAD

**Issue by manopapad** [StanfordLegion/legion#431]
Pull https://github.com/stanfordhpccenter/soleil-x/tree/bug/segfault-wo-freeze and follow the setup instructions for sapling. Then run:
```
USE_CUDA=0 ./soleil.sh -m ../testcases/hit_to_openchannel/upper.json
```
The application will crash with a segfault and exit, even though we are running with `LEGION_FREEZE_ON_ERROR=1`:
```
*** Caught a signal: SIGTERM(15) on node 1/2
--------------------------------------------------------------------------
mpiexec noticed that process rank 0 with PID 19115 on node n0002 exited on signal 11 (Segmentation fault).
--------------------------------------------------------------------------
```
This is a CPU-only run on 2 nodes. You can add debug output to `soleil_mapper.cc` to verify that the runtime has started. The application doesn't make any progress, so this is not an issue of segfaulting after the runtime had shut down.

**Comment by lightsighter** [comment]
Are you passing all your environment variables to mpirun with the '-x' flag?

**Comment by manopapad** [comment]
Yes https://github.com/stanfordhpccenter/soleil-x/blob/bug/segfault-wo-freeze/src/run.sh#L121-L127

**Comment by manopapad** [comment]
Sean posited that this has something to do with the stack getting corrupted, and that caused both the original segfault, and also a segfault inside the signal handler.

**Comment by manopapad** [comment]
I fixed my original issue. The `workDual` task requires ~3MB of stack space, and I was only setting `-ll:ostack` (which controls the stack space allocated to OpenMP threads) but not `-ll:stacksize` (which controls the same value for regular CPU threads). This caused Soleil to run out of stack space when running with just CPU processors. This should have been immediately obvious to me if I had bothered to `disas $<workDual>` at crash time:

```
Dump of assembler code for function $<workDual>:
   0x00000000012254a0 <+0>:    push   %rbp
   0x00000000012254a1 <+1>:    mov    %rsp,%rbp
   0x00000000012254a4 <+4>:    push   %r15
   0x00000000012254a6 <+6>:    push   %r14
   0x00000000012254a8 <+8>:    push   %r13
   0x00000000012254aa <+10>:    push   %r12
   0x00000000012254ac <+12>:    push   %rbx
   0x00000000012254ad <+13>:    sub    $0x328a98,%rsp
=> 0x00000000012254b4 <+20>:    mov    %rdi,-0x328270(%rbp)
   0x00000000012254bb <+27>:    mov    %rsi,-0x328268(%rbp)
   0x00000000012254c2 <+34>:    mov    %edx,-0x32825c(%rbp)
   0x00000000012254c8 <+40>:    mov    %rcx,-0x328258(%rbp)
```

In this situation the signal handler is invoked but immediately crashes, because `$esp` is pointing outside of mapped memory. Sean would like the signal handler to also catch this sort of error, and to accomplish this the signal handlers would need to be using an alternate stack. I'm keeping this issue open (and changing the name) to track work on this.

**Comment by streichler** [comment]
Assigning this to me.  As @manopapad suggests, I want to look at using `sigaltstack()` to catch segfaults in a more robust way.

**Comment by streichler** [comment]
This should be fixed with commit eb0c4131dbc61627d60e40f317c885fbc16b52d1.  @manopapad, do you want to test it out, or just take my word for it?

**Comment by manopapad** [comment]
I have tested the fix and can confirm that Legion now freezes successfully even when we overflow the stack.
